File: src/js/semantic-cutout.js
const STRATEGIES = new Set(['foreground', 'semantic']);

function clamp(value, min, max) {
  return Math.min(max, Math.max(min, Number(value) || 0));
}

function normalizeModelQuery(value) {
  return String(value || '').trim().toLowerCase().replace(/\.+$/, '').slice(0, 80);
}

function normalizeRegion(region, index, query) {
  const raw = Array.isArray(region?.bbox) ? region.bbox : [];
  const x = clamp(raw[0], 0, 1);
  const y = clamp(raw[1], 0, 1);
  const width = clamp(raw[2], 0, 1 - x);
  const height = clamp(raw[3], 0, 1 - y);
  const rawOrigin = String(region?.origin || '');
  const origin = ['automatic', 'automatic-review'].includes(rawOrigin) ? rawOrigin : 'manual';
  const normalized = {
    id: String(region?.id || `target-${index + 1}`),
    label: String(region?.label || query || `目标 ${index + 1}`).trim(),
    bbox: [x, y, width, height].map((value) => Number(value.toFixed(6))),
    origin,
  };
  if (['automatic', 'automatic-review'].includes(normalized.origin)) {
    normalized.confidence = Number(clamp(region?.confidence, 0, 1).toFixed(4));
  }
  return normalized;
}

function normalizeMaskEdit(edit) {
  const mode = edit?.mode === 'include' ? 'include' : edit?.mode === 'exclude' ? 'exclude' : '';
  const points = Array.from(edit?.points || [], (point) => [
    Number(clamp(point?.[0], 0, 1).toFixed(6)),
    Number(clamp(point?.[1], 0, 1).toFixed(6)),
  ]).filter((point) => point.every(Number.isFinite));
  if (!mode || !points.length) return null;
  return {
    mode,
    points,
    radius: Number(clamp(edit?.radius ?? 0.018, 0.003, 0.1).toFixed(6)),
  };
}

export function createSemanticCutoutState(raw = {}) {
  const strategy = STRATEGIES.has(raw?.strategy) ? raw.strategy : 'foreground';
  const query = String(raw?.query || '').trim().slice(0, 80);
  const modelQuery = normalizeModelQuery(raw?.model_query);
  const modelQueryOverride = normalizeModelQuery(raw?.model_query_override);
  const targetCount = Math.round(clamp(raw?.target_count ?? 1, 1, 8));
  const regions = Array.from(raw?.regions || [], (region, index) => normalizeRegion(region, index, query))
    .filter((region) => region.bbox[2] > 0 && region.bbox[3] > 0);
  const maskEdits = Array.from(raw?.mask_edits || [], normalizeMaskEdit).filter(Boolean);
  return {
    strategy,
    query,
    model_query: modelQuery,
    model_query_override: modelQueryOverride,
    target_count: targetCount,
    source_asset_id: String(raw?.source_asset_id || ''),
    status: raw?.status === 'confirmed' ? 'confirmed' : 'draft',
    method: ['manual-box', 'model-candidate-confirmed', 'model-assisted-confirmed'].includes(raw?.method)
      ? raw.method
      : 'manual-box',
    digest: String(raw?.digest || ''),
    regions,
    mask_edits: maskEdits,
  };
}

export function updateSemanticCutoutState(current, patch = {}) {
  const before = createSemanticCutoutState(current);
  const next = createSemanticCutoutState({ ...before, ...patch });
  const invalidatesConfirmation = (
    next.query !== before.query
    || next.model_query_override !== before.model_query_override
    || next.target_count !== before.target_count
    || (
      Object.prototype.hasOwnProperty.call(patch, 'source_asset_id')
      && next.source_asset_id !== before.source_asset_id
    )
    || (
      Object.prototype.hasOwnProperty.call(patch, 'mask_edits')
      && JSON.stringify(next.mask_edits) !== JSON.stringify(before.mask_edits)
    )
  );
  const carriesFreshConfirmation = (
    patch?.status === 'confirmed'
    && Boolean(patch?.digest)
    && Array.isArray(patch?.regions)
  );
  if (invalidatesConfirmation && !carriesFreshConfirmation) {
    next.status = 'draft';
    next.model_query = '';
    next.digest = '';
    next.source_asset_id = '';
    next.regions = [];
    next.mask_edits = [];
  }
  return next;
}

export function semanticCutoutReadiness(rawState, selectedAssetIds = []) {
  const state = createSemanticCutoutState(rawState);
  const sourceIds = Array.from(selectedAssetIds || [], String);
  if (state.strategy === 'foreground') {
    return sourceIds.length
      ? { action: 'submit', ready: true, message: `${sourceIds.length} 张素材 · 本地分离全部前景` }
      : { action: 'select-source', ready: false, message: '从抠图素材中选择后可入队' };
  }
  if (!sourceIds.length) return { action: 'select-source', ready: false, message: '先选择 1 张需要选物的图片' };
  if (sourceIds.length !== 1) return { action: 'single-source', ready: false, message: '智能选物首版每次确认 1 张；请只保留一张选中素材' };
  if (!state.query) return { action: 'describe', ready: false, message: '填写要保留的物体名称，例如“汉堡”' };
  const confirmed = (
    state.status === 'confirmed'
    && Boolean(state.digest)
    && state.source_asset_id === sourceIds[0]
    && state.regions.length === state.target_count
  );
  if (!confirmed) {
    return {
      action: 'confirm',
      ready: false,
      message: `请在原图上确认 ${state.target_count} 个“${state.query}”`,
    };
  }
  return {
    action: 'submit',
    ready: true,
    message: `已确认 ${state.target_count} 个“${state.query}” · 本地执行，不消耗生图额度`,
  };
}

export function semanticCutoutPayload(rawState, selectedAssetIds = []) {
  const state = createSemanticCutoutState(rawState);
  if (state.strategy === 'foreground') return { strategy: 'foreground' };
  const sourceId = String(selectedAssetIds?.[0] || state.source_asset_id || '');
  return {
    strategy: 'semantic',
    query: state.query,
    model_query: state.model_query,
    target_count: state.target_count,
    sources: {
      [sourceId]: {
        source_asset_id: sourceId,
        status: state.status,
        method: state.method,
        digest: state.digest,
        regions: state.regions.map((region) => ({ ...region, bbox: [...region.bbox] })),
        mask_edits: state.mask_edits.map((edit) => ({
          ...edit,
          points: edit.points.map((point) => [...point]),
        })),
      },
    },
  };
}

export function semanticCutoutStageCopy(stage) {
  return {
    recognition: '目标名称无法识别，请改写名称或手动框选',
    selection: '目标数量或选区不一致，请重新确认',
    segmentation: '框选区域内没有得到有效前景，请扩大选区',
    edge: '边缘精修失败，请改用快速去背景或重新框选',
  }[stage] || '智能选物暂时无法继续，请检查目标确认状态';
}

export function semanticGroundingPresentation(rawPreview = {}) {
  const grounding = rawPreview?.grounding || {};
  const status = String(rawPreview?.candidate_status || grounding?.status || 'unavailable');
  const fallback = {
    candidates: '本地模型已给出候选，请逐个检查；确认前不会开始抠图',
    low_confidence: '候选置信度不足，请修正选区或补充框选',
    no_match: '没有找到可靠候选，请手动框选目标',
    query_unmapped: '当前名称还没有离线识别词；可填写英文识别词或手动框选',
    unavailable: '当前未配置本地目标定位模型，请手动框选',
    failed: '本地自动定位失败，请手动框选；当前图片仍可继续处理',
    manual_regions: '请检查手动选区后确认',
  }[status] || '请检查目标选区后确认';
  const tone = {
    candidates: 'candidate',
    low_confidence: 'warning',
    failed: 'error',
  }[status] || 'manual';
  return {
    status,
    tone,
    message: String(rawPreview?.message || grounding?.message || fallback),
  };
}

File: python/semantic_cutout.py
from __future__ import annotations

import hashlib
import json
import math
from typing import Any, Iterable

from PIL import Image, ImageChops, ImageDraw


class SemanticCutoutError(ValueError):
    def __init__(self, code: str, message: str, *, stage: str) -> None:
        super().__init__(message)
        self.code = code
        self.message = message
        self.stage = stage


def _query(value: Any) -> str:
    query = " ".join(str(value or "").strip().split())[:80]
    if not query:
        raise SemanticCutoutError(
            "SEMANTIC_QUERY_REQUIRED",
            "请先填写要保留的物体名称",
            stage="recognition",
        )
    return query


def _target_count(value: Any) -> int:
    try:
        count = int(value)
    except (TypeError, ValueError) as exc:
        raise SemanticCutoutError(
            "SEMANTIC_TARGET_COUNT_INVALID",
            "目标数量必须是 1 到 8 的整数",
            stage="selection",
        ) from exc
    if count < 1 or count > 8:
        raise SemanticCutoutError(
            "SEMANTIC_TARGET_COUNT_INVALID",
            "目标数量必须是 1 到 8 的整数",
            stage="selection",
        )
    return count


def _model_query(value: Any) -> str:
    model_query = " ".join(str(value or "").strip().split()).lower().rstrip(".")[:80]
    if any("\u3400" <= char <= "\u9fff" or "\uf900" <= char <= "\ufaff" for char in model_query):
        raise SemanticCutoutError(
            "SEMANTIC_MODEL_QUERY_INVALID",
            "模型识别词必须使用英文；可清空后重新自动映射",
            stage="recognition",
        )
    return model_query


def _number(value: Any, field: str) -> float:
    try:
        result = float(value)
    except (TypeError, ValueError) as exc:
        raise SemanticCutoutError(
            "SEMANTIC_REGION_INVALID",
            f"选区 {field} 不是有效数字",
            stage="selection",
        ) from exc
    if not math.isfinite(result):
        raise SemanticCutoutError(
            "SEMANTIC_REGION_INVALID",
            f"选区 {field} 不是有限数字",
            stage="selection",
        )
    return result


def normalize_regions(regions: Any, query: str) -> list[dict[str, Any]]:
    if not isinstance(regions, list):
        raise SemanticCutoutError(
            "SEMANTIC_REGIONS_REQUIRED",
            "请在原图上框选要保留的目标",
            stage="selection",
        )
    normalized = []
    for index, region in enumerate(regions):
        if not isinstance(region, dict) or not isinstance(region.get("bbox"), (list, tuple)):
            raise SemanticCutoutError(
                "SEMANTIC_REGION_INVALID",
                f"第 {index + 1} 个选区缺少边界",
                stage="selection",
            )
        bbox = region["bbox"]
        if len(bbox) != 4:
            raise SemanticCutoutError(
                "SEMANTIC_REGION_INVALID",
                f"第 {index + 1} 个选区边界必须包含 x、y、宽、高",
                stage="selection",
            )
        x, y, width, height = (
            _number(bbox[0], "x"),
            _number(bbox[1], "y"),
            _number(bbox[2], "width"),
            _number(bbox[3], "height"),
        )
        if x < 0 or y < 0 or width <= 0 or height <= 0 or x + width > 1 or y + height > 1:
            raise SemanticCutoutError(
                "SEMANTIC_REGION_OUT_OF_BOUNDS",
                f"第 {index + 1} 个选区超出图片范围",
                stage="selection",
            )
        if width < 0.01 or height < 0.01:
            raise SemanticCutoutError(
                "SEMANTIC_REGION_TOO_SMALL",
                f"第 {index + 1} 个选区太小，请扩大后重试",
                stage="selection",
            )
        normalized_region = {
            "id": str(region.get("id") or f"target-{index + 1}")[:80],
            "label": str(region.get("label") or query).strip()[:80] or query,
            "bbox": [round(value, 6) for value in (x, y, width, height)],
            "origin": "automatic" if region.get("origin") == "automatic" else "manual",
        }
        if normalized_region["origin"] == "automatic":
            try:
                confidence = float(region.get("confidence", 0))
            except (TypeError, ValueError):
                confidence = 0.0
            normalized_region["confidence"] = round(max(0.0, min(1.0, confidence)), 4)
        normalized.append(normalized_region)
    return normalized


def _selection_digest(
    source_asset_id: str,
    query: str,
    target_count: int,
    method: str,
    regions: list[dict[str, Any]],
    model_query: str = "",
) -> str:
    payload = {
        "source_asset_id": source_asset_id,
        "query": query,
        "target_count": target_count,
        "method": method,
        "regions": regions,
    }
    if model_query:
        payload["model_query"] = model_query
    canonical = json.dumps(payload, ensure_ascii=False, sort_keys=True, separators=(",", ":"))
    return "sha256:" + hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def build_confirmed_selection(
    *,
    source_asset_id: str,
    query: Any,
    target_count: Any,
    regions: Any,
    model_query: Any = "",
) -> dict[str, Any]:
    asset_id = str(source_asset_id or "").strip()
    if not asset_id:
        raise SemanticCutoutError(
            "SEMANTIC_SOURCE_REQUIRED",
            "请先选择一张源图片",
            stage="selection",
        )
    normalized_query = _query(query)
    normalized_model_query = _model_query(model_query)
    count = _target_count(target_count)
    normalized_regions = normalize_regions(regions, normalized_query)
    if len(normalized_regions) != count:
        raise SemanticCutoutError(
            "SEMANTIC_TARGET_COUNT_MISMATCH",
            f"要求保留 {count} 个目标，但当前确认了 {len(normalized_regions)} 个",
            stage="selection",
        )
    automatic_count = sum(region.get("origin") == "automatic" for region in normalized_regions)
    method = (
        "model-candidate-confirmed"
        if automatic_count == len(normalized_regions)
        else "model-assisted-confirmed"
        if automatic_count
        else "manual-box"
    )
    source_plan = {
        "source_asset_id": asset_id,
        "status": "confirmed",
        "method": method,
        "regions": normalized_regions,
    }
    source_plan["digest"] = _selection_digest(
        asset_id,
        normalized_query,
        count,
        method,
        normalized_regions,
        normalized_model_query,
    )
    return {
        "strategy": "semantic",
        "query": normalized_query,
        "model_query": normalized_model_query,
        "target_count": count,
        "sources": {asset_id: source_plan},
    }


def normalize_cutout_selection(value: Any) -> dict[str, Any]:
    if value in (None, "", {}):
        return {"strategy": "foreground"}
    if not isinstance(value, dict):
        raise SemanticCutoutError(
            "CUTOUT_SELECTION_INVALID",
            "抠图选择参数必须是对象",
            stage="selection",
        )
    strategy = str(value.get("strategy") or "foreground").strip().lower()
    if strategy == "foreground":
        return {"strategy": "foreground"}
    if strategy != "semantic":
        raise SemanticCutoutError(
            "CUTOUT_STRATEGY_UNSUPPORTED",
            "不支持的抠图策略",
            stage="selection",
        )
    query = _query(value.get("query"))
    model_query = _model_query(value.get("model_query"))
    count = _target_count(value.get("target_count"))
    raw_sources = value.get("sources")
    if not isinstance(raw_sources, dict):
        raw_sources = {}
    sources: dict[str, dict[str, Any]] = {}
    for key, raw_plan in raw_sources.items():
        if not isinstance(raw_plan, dict):
            raise SemanticCutoutError(
                "SEMANTIC_CONFIRMATION_INVALID",
                "目标确认记录无效，请重新确认",
                stage="selection",
            )
        asset_id = str(raw_plan.get("source_asset_id") or key).strip()
        if asset_id != str(key):
            raise SemanticCutoutError(
                "SEMANTIC_SOURCE_MISMATCH",
                "目标确认记录与源图片不一致，请重新确认",
                stage="selection",
            )
        method = str(raw_plan.get("method") or "")
        if raw_plan.get("status") != "confirmed" or method not in {
            "manual-box", "model-candidate-confirmed", "model-assisted-confirmed",
        }:
            raise SemanticCutoutError(
                "SEMANTIC_CONFIRMATION_REQUIRED",
                "智能选物必须先逐张确认目标",
                stage="selection",
            )
        regions = normalize_regions(raw_plan.get("regions"), query)
        if len(regions) != count:
            raise SemanticCutoutError(
                "SEMANTIC_TARGET_COUNT_MISMATCH",
                f"要求保留 {count} 个目标，但当前确认了 {len(regions)} 个",
                stage="selection",
            )
        expected = _selection_digest(asset_id, query, count, method, regions, model_query)
        if str(raw_plan.get("digest") or "") != expected:
            raise SemanticCutoutError(
                "SEMANTIC_CONFIRMATION_STALE",
                "目标名称、数量或选区已经变化，请重新确认",
                stage="selection",
            )
        sources[asset_id] = {
            "source_asset_id": asset_id,
            "status": "confirmed",
            "method": method,
            "digest": expected,
            "regions": regions,
        }
    return {
        "strategy": "semantic",
        "query": query,
        "model_query": model_query,
        "target_count": count,
        "sources": sources,
    }


def validate_selection_sources(selection: dict[str, Any], source_asset_ids: Iterable[str]) -> None:
    if selection.get("strategy") != "semantic":
        return
    expected = [str(asset_id) for asset_id in source_asset_ids]
    actual = list((selection.get("sources") or {}).keys())
    missing = [asset_id for asset_id in expected if asset_id not in actual]
    extra = [asset_id for asset_id in actual if asset_id not in expected]
    if missing or extra:
        raise SemanticCutoutError(
            "SEMANTIC_CONFIRMATION_REQUIRED",
            "智能选物必须对本次每张源图逐张确认，且不能沿用其他图片的选区",
            stage="selection",
        )


def apply_confirmed_regions(image: Image.Image, regions: Any) -> Image.Image:
    normalized = normalize_regions(regions, "目标")
    if not normalized:
        raise SemanticCutoutError(
            "SEMANTIC_REGIONS_REQUIRED",
            "没有可执行的目标选区",
            stage="selection",
        )
    rgba = image.convert("RGBA")
    mask = Image.new("L", rgba.size, 0)
    draw = ImageDraw.Draw(mask)
    width, height = rgba.size
    for region in normalized:
        x, y, box_width, box_height = region["bbox"]
        left = max(0, min(width, math.floor(x * width)))
        top = max(0, min(height, math.floor(y * height)))
        right = max(left + 1, min(width, math.ceil((x + box_width) * width)))
        bottom = max(top + 1, min(height, math.ceil((y + box_height) * height)))
        draw.rectangle((left, top, right - 1, bottom - 1), fill=255)
    alpha = ImageChops.multiply(rgba.getchannel("A"), mask)
    if alpha.getbbox() is None:
        raise SemanticCutoutError(
            "SEMANTIC_SEGMENTATION_EMPTY",
            "框选区域内没有得到有效前景，请扩大选区后重试",
            stage="segmentation",
        )
    rgba.putalpha(alpha)
    return rgba
